dali/kernels/imgproc/convolution/separable_convolution_cpu.h

```diff
@@ -0,0 +1,202 @@
+ *
+ * If `Out` is same as `W` the intermediate stages are written to out,
+ * otherwise the temporary buffer of `W` is required and allocated in scratchpad
+ *
+ * `W` is currently assumed to be float
+ *
+ * `windows` and `scales` are specified per axis.
+ *
+ * Specialized for 1, 2 or 3 axes, to not go overboard with TMP for generic solutions
+ *
+ * TODO(klecki): For more dimension, fusing a permute step when writing the result
+ * could allow for processing all steps with innermost, contiguous dimension.
+ * For example DHWC->DWHC->HWDC->DHWC, while applying convolutions for W, H, D respectively.
+ */
+template <typename Out, typename In, typename W, int ndim, bool has_channels = true>
+struct SeparableConvolutionCpu;
+
+template <typename Out, typename W, int ndim>
+std::enable_if_t<!std::is_same<Out, W>::value, TensorView<StorageCPU, W, ndim>> GetInterStageView(
+    const TensorView<StorageCPU, Out, ndim>& out, W* scratch) {
+  return {scratch, out.shape};
+}
+
+template <typename W, int ndim>
+TensorView<StorageCPU, W, ndim> GetInterStageView(const TensorView<StorageCPU, W, ndim>& out,
+                                                  W* scratch) {
+  return out;
+}
+
+template <typename Out, typename In, typename W, int axes, bool has_channels>
+struct SeparableConvolutionCpuImpl;
+
+
+template <typename Out, typename In, typename W, bool has_channels>
+struct SeparableConvolutionCpuImpl<Out, In, W, 1, has_channels> {
+  static constexpr int axes = 1;
+  static constexpr int ndim = has_channels ? 2 : 1;
+
+  KernelRequirements Setup(KernelContext& ctx, const TensorShape<ndim>& in_shape,
+                           const std::array<int, axes>& window_sizes) {
+    KernelRequirements req;
+    req.output_shapes.push_back(uniform_list_shape<ndim>(1, in_shape));
+
+    auto req_conv = conv_.Setup(ctx, in_shape, window_sizes[0]);
+    req.AddInputSet(req_conv, false);
+
+    return req;
+  }
+
+  void Run(KernelContext& ctx, const TensorView<StorageCPU, Out, ndim> out,
+           const TensorView<StorageCPU, const In, ndim>& in,
+           const std::array<TensorView<StorageCPU, const W, 1>, axes>& windows,
+           const std::array<W, axes>& scales = uniform_array<axes, W>(1.f)) {
+    conv_.Run(ctx, out, in, windows[0], scales[0]);
+  }
+
+  ConvolutionCpu<Out, In, W, ndim, 0, has_channels> conv_;
+  static_assert(std::is_same<W, float>::value,
+                "Only floats as intermediate values are currently supported.");
+};
+
+template <typename Out, typename In, typename W, bool has_channels>
+struct SeparableConvolutionCpuImpl<Out, In, W, 2, has_channels> {
+  static constexpr int axes = 2;
+  static constexpr int ndim = has_channels ? 3 : 2;
+
+  KernelRequirements Setup(KernelContext& ctx, const TensorShape<ndim>& in_shape,
+                           const std::array<int, axes>& window_sizes) {
+    KernelRequirements req;
+
+    ScratchpadEstimator se;
+    se.add<W>(AllocType::Host, volume(in_shape));
```

Comment:
Why do you want the intermedate image to be of type W? What if you have `float` input/output and integer weights?

Reply:
Also, you don't need this buffer when your intermediate element type is the same as Out.

Reply:
> Why do you want the intermediate image to be of type W? What if you have float input/output and integer weights?

For now I assume that `W` is float. It can be generalized as well, and we can parametrize every possible step here with configurable type. Do you want me to do it?

> Also, you don't need this buffer when your intermediate element type is the same as Out.

Yes, I don't need it, but the in-place for first step will be slower, and for everything other than W==Out it's still needed so I just opted for that.

Reply:
As for the intermediate data, maybe it would indeed be better to just store the result of arithmetic operation. Will do that.